### PR TITLE
Store invalid blocks

### PR DESCRIPTION
**Pull request by sadiq1971**

### CODE REVIEW

File: packages/lodestar/src/chain/chain.ts [hunk 1]
```diff
@@ -165,10 +165,17 @@ export class BeaconChain extends EventEmitter implements IBeaconChain {
     const preFinalizedEpoch = state.finalizedEpoch;
     const preJustifiedEpoch = state.currentJustifiedEpoch;
     // Run the state transition
-    const newState = stateTransition(this.config, state, block, true);
+    let newState: BeaconState;
+    try {
+      newState = stateTransition(this.config, state, block, true);
+    }catch (e) {
+      // store block root in db and terminate
```
Comment (mpetrunic) [inline packages/lodestar/src/chain/chain.ts:172]: add logger warn or error that you caught bad block with some info like why did it fail

File: packages/lodestar/src/sync/regular.ts [hunk 2]
```diff
@@ -32,9 +32,13 @@ export class RegularSync {
   }
 
   public async receiveBlock(block: BeaconBlock): Promise<void> {
-    // TODO: skip block if its a known bad block
-    // skip block if it already exists
     const root = hashTreeRoot(block, this.config.types.BeaconBlock);
+
+    // skip block if its a known bad block
+    if (await this.db.isBadBlockRoot(root)){
+      return ;
```
Comment (mpetrunic) [inline packages/lodestar/src/sync/regular.ts:40]: add log of what happened, probably warn

**Review by mpetrunic — CHANGES_REQUESTED** [review]

File: packages/lodestar/src/db/api/beacon/beacon.ts [hunk 3]
```diff
@@ -320,4 +320,17 @@ export class BeaconDB extends DatabaseService implements IBeaconDb {
     );
   }
 
+  public async setBadBlockRoot(root: bytes32): Promise<void> {
+    return await this.db.put(encodeKey(Bucket.invalidBlock, root),
+      serialize(root, this.config.types.bytes32));
+  }
+  public async isBadBlockRoot(root: bytes32): Promise<boolean> {
```
Comment (GregTheGreek) [inline packages/lodestar/src/db/api/beacon/beacon.ts:330]: add whitespace

**Review by GregTheGreek — CHANGES_REQUESTED** [review]

File: packages/lodestar/src/chain/chain.ts [hunk 4]
```diff
@@ -165,10 +165,18 @@ export class BeaconChain extends EventEmitter implements IBeaconChain {
...
+      const blockRoot = hashTreeRoot(block, this.config.types.BeaconBlock);
+      await this.db.setBadBlockRoot(blockRoot);
+      this.logger.warn( "bad block" + e.message + '\n' + e.stack);
```
Comment (mpetrunic) [inline packages/lodestar/src/chain/chain.ts]: - put as second line of catch statement
- add blockroot to log
- missing whitespace after "bad block"
- don't put stack there, you could print stack as a debug log or something

File: packages/lodestar/src/sync/regular.ts [hunk 5]
```diff
@@ -32,9 +32,14 @@ export class RegularSync {
   }
 
   public async receiveBlock(block: BeaconBlock): Promise<void> {
-    // TODO: skip block if its a known bad block
-    // skip block if it already exists
     const root = hashTreeRoot(block, this.config.types.BeaconBlock);
+
+    // skip block if its a known bad block
+    if (await this.db.isBadBlockRoot(root)){
+      this.logger.warn("Found bad block, skipping");
```
Comment (mpetrunic) [inline packages/lodestar/src/sync/regular.ts]: add block root to  log, this is meaningless. Format it to says "received" instead of "found"

**Review by mpetrunic — CHANGES_REQUESTED** [review]

**Review by mpetrunic — APPROVED** [review]

**Review by GregTheGreek — APPROVED** [review]

**Pull request merged**